File: res/update.mjs
import fs, { writeFileSync } from 'fs';
import { exit } from 'process';

let project = '';

try {
	project = fs.readFileSync('./lambda_version.hpp').toString();
} catch (error) {
	console.error('Run this script from project root, not wherever you currently are');
	exit(1);
}

const dataImported = {
	version: project.match(/\#define\s+LAMBDA_VERSION\s+\"[\d\.]+\"/)?.[0].match(/"[\d\.]+\"$/)?.[0]?.replace(/\"/g, ''),
	product: project.match(/\#define\s+LAMBDA_PRODUCT\s+\".+\"/)?.[0].match(/".+\"$/)?.[0]?.replace(/\"/g, '')
};

console.log(dataImported);

const resourceTemplate = fs.readFileSync('./res/template.rc').toString();

let outputcontent = resourceTemplate;

const replaces = [
	{
		variable: 'version_comma',
		data: dataImported.version.replace(/\./g, ',')
	},
	{
		variable: 'version',
		data: dataImported.version
	},
	{
		variable: 'released',
		data: new Date().getFullYear()
	},
	{
		variable: 'product',
		data: dataImported.product
	}
];

replaces.forEach(item => outputcontent = outputcontent.replace(new RegExp(`\\$\\{${item.variable}\\}`, 'g'), item.data));

writeFileSync('./res/lambda.rc', outputcontent);

File: res/update.py
import re
import datetime
import os
import sys


# check for correct directory
if not os.getcwd().endswith("res"):
	print("The paths for this script are relative to `res/` directory, so `cd res` and run this script over there")
	sys.exit(0)


# read contents of lambda_version.hpp
versionHeaderFile = open('../lambda_version.hpp', 'r')
versionHeaderContent = versionHeaderFile.read()
versionHeaderFile.close()


# replace values
insertVariables = [
	{  "var": "released", "data":  str(datetime.date.today().year) }
]


# lambda version
versionLine = re.search("\#define\s+LAMBDA_VERSION\s+\"[\d\.]+\"", versionHeaderContent)
versionMatch = re.search("\"[\d\.]+\"$", versionLine.group()).group()
version = re.sub(r'\"', '', versionMatch)

insertVariables.append({ "var": "version", "data":  version })
insertVariables.append({ "var": "version_comma", "data":  re.sub(r'\.', ',', version) })


# lambda name
productLine = re.search("\#define\s+LAMBDA_PRODUCT\s+\".+\"", versionHeaderContent)
productMatch = re.search("\".+\"$", productLine.group()).group()
productName = re.sub(r'\"', '', productMatch)

insertVariables.append({ "var": "product", "data":  productName })


# generate rc file from the template
templateFile = open('template.rc', 'r')
template = templateFile.read()
templateFile.close()

for item in insertVariables:
	template = re.sub(r"\$\{" + item["var"] + "\}", item["data"], template)

# write results
rcFile = open("lambda.rc", 'w')
rcFile.write(template)
rcFile.close()

print(">>> rc file generated")
